arangod/VocBase/vocbase.cpp

```diff
@@ -819,45 +805,41 @@ arangodb::Result TRI_vocbase_t::loadCollection(
   }
 
   // read lock
-  // check if the collection is already loaded
   READ_LOCKER_EVENTUAL(locker, collection.statusLock());
-  TRI_vocbase_col_status_e status = collection.status();
 
-  if (status == TRI_VOC_COL_STATUS_LOADED) {
-    // DO NOT release the lock
-    locker.steal();
-    return {};
-  }
-
-  if (status == TRI_VOC_COL_STATUS_DELETED) {
+  if (collection.deleted()) {
     return {TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND,
             std::string("collection '") + collection.name() + "' not found"};
   }
 
-  if (status == TRI_VOC_COL_STATUS_CORRUPTED) {
-    return {TRI_ERROR_ARANGO_CORRUPTED_COLLECTION};
-  }
-
-  TRI_ASSERT(false);
-  return {TRI_ERROR_INTERNAL, "unknwon collection status"};
+  // DO NOT release the lock
+  locker.steal();
+  return {};
 }
 
 /// @brief drops a collection, worker function
-ErrorCode TRI_vocbase_t::dropCollectionWorker(
-    arangodb::LogicalCollection* collection, DropState& state, double timeout) {
-  state = DROP_EXIT;
-  std::string const colName(collection->name());
+arangodb::Result TRI_vocbase_t::dropCollectionWorker(
+    arangodb::LogicalCollection& collection) {
+  std::string const colName(collection.name());
+  std::string const& dbName = _info.getName();
 
-  StorageEngine& engine = server().getFeature<EngineSelectorFeature>().engine();
-  engine.prepareDropCollection(*this, *collection);
+  arangodb::Result res;
 
-  double endTime = TRI_microtime() + timeout;
+  auto guard = scopeGuard([&res, &colName, &dbName]() noexcept {
+    try {
+      events::DropCollection(dbName, colName, res.errorNumber());
```

Comment:
Previously there were cases where we returned from the function _without_ calling events::DopCollection (e.g. TRI_ERROR_SHUTTING_DOWN). With the scopeGuard this behavior changes. Is this intentional?

Reply:
Yes!